servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyConnection.java

```diff
@@ -459,6 +459,21 @@ public FlushStrategy defaultFlushStrategy() {
          */
         void channelWritable();
 
+        /**
+         * Notifies the end of payload was observed at the transport (not necessarily flushed).
+         *
+         * <p>
+         * This helps coordinate state between the writer and the codec/transport. Even if the writing source may not
+         * have completed from a Reactive Streams perspective, it means that as far as the codec and transport is
+         * concerned it considers the write is complete and no further writes can happen. Closing of the transport after
+         * receiving this event should not be considered a write error since the final buffer was written and
+         * potentially flushed. A failed write promise after observing this event however indicates the end of payload
+         * marker may not have been written correctly and should result in an error.
+         * <p>
+         * This is event is expected be called from the eventloop.
+         */
+        void protocolPayloadComplete();
```

Comment:
I think we can leverage pre-existing concepts here instead of introducing `protcol` and `payload` knowledge inside `WritableListener`. What we are after here is essentially "graceful closure" which in this case means wait for enqueued writes to finish before closing the listener.

I would suggest renaming this method to `closeGracefully()` and renaming existing `channelClosed(Throwable)` to `close(Throwable)` so that we clearly can define behaviors for abrupt and graceful closures without introducing the "protocol concept" here. If you get around unifying `channelClosedOutbound()` and `channelClosed()` as mentioned above then we can also remove `channelClosedOutbound()`.

Reply:
OK I'll try